fenicsadapter/fenicsadapter.py

```diff
@@ -159,46 +164,86 @@ def create_coupling_neumann_boundary_condition(self, test_functions):
+        
+        This means:
+        * either, the checkpoint self._u_cp is assigned to u_n to repeat the iteration,
+        * or u_n+1 is assigned to u_n and the checkpoint is updated correspondingly.
+        
+        :param write_function: a FEniCS function being sent to the other participant as boundary condition at the coupling interface
+        :param u_np1: new value of FEniCS solution u_n+1 at time t_n+1 = t+dt
+        :param u_n: old value of FEniCS solution u_n at time t_n = t; updated via call-by-reference
+        :param t: current time t_n for timestep n
+        :param dt: timestep size dt = t_n+1 - t_n
+        :param n: current timestep
+        :return: return starting time t and timestep n for next FEniCS solver iteration. u_n is updated by advance correspondingly. 
+        """
+        # sample write data at interface
         x_vert, y_vert = self.extract_coupling_boundary_coordinates()
         self._write_data = self.convert_fenics_to_precice(write_function, self._mesh_fenics, self._coupling_subdomain)
+
+        # communication
         self._interface.writeBlockScalarData(self._write_data_id, self._n_vertices, self._vertex_ids, self._write_data)
         self._interface.advance(dt)
         self._interface.readBlockScalarData(self._read_data_id, self._n_vertices, self._vertex_ids, self._read_data)
+
+        # update boundary condition with read data
         self._coupling_bc_expression.update_boundary_data(self._read_data, x_vert, y_vert)
+
+        success = False
+
+        # checkpointing
         if self._interface.isActionRequired(PySolverInterface.PyActionReadIterationCheckpoint()):
-            # We do not require proper checkpointing for the FEniCS examples we are currently dealing with.
-            # do nothing
+            # continue FEniCS computation from checkpoint
+            # todo we might want to put reading the checkpoint into a function (duplicate code. compare to below)
+            u_n.assign(self._u_cp)  # set u_n to value of checkpoint
+            t = self._t_cp
+            n = self._n_cp
             self._interface.fulfilledAction(PySolverInterface.PyActionReadIterationCheckpoint())
-            return False
-        else:
-            return True
 
-    def initialize(self, coupling_subdomain, mesh, read_field, write_field):
+        if self._interface.isActionRequired(PySolverInterface.PyActionWriteIterationCheckpoint()):
+            # continue FEniCS computation with u_np1
+            # update checkpoint
+            self._u_cp.assign(u_np1)
+            assert (np.isclose(t + dt, self._t_cp + dt))
+            self._t_cp = t + dt
+            assert (np.isclose(n + 1, self._n_cp + 1))
+            self._n_cp = n + 1
+            # todo we might want to put reading the checkpoint into a function (duplicate code. compare to above)
+            u_n.assign(self._u_cp)  # set u_n to value of (updated)checkpoint
```

Comment:
I think this one should go outside the if-statement. `WriteIterationCheckpoint` is only fired if a new timestep is entered for the first time. But if FEniCS subcycles `ReadIterationCheckpoint` and `WriteIterationCheckpoint` are false. Still, you want that `u_n.assign(u_np1)` and that time and n increases

Reply:
If we want to support subcycling, I think we have to pull lines 216-218 out of this if loop and write something like this:
```
u_n.assign(u_np1)
t = t+dt
n=n+1
```
Be careful: If a checkpoint is read, these lines should **not** be executed! Maybe put them in the `else` of `if self._interface.isActionRequired(PySolverInterface.PyActionReadIterationCheckpoint()):` (line 199)

Reply:
See https://github.com/precice/fenics-adapter/pull/19/commits/66fd9f98992308a99cfa094b83629478abfe52b8 for a resolution. However, we still have an issue with non-matching subcycling #20

Reply:
We decided to ignore #20 for this PR